Core Metadata v2.1

donmccurdy opened the pull request:
Language and clarity pass on the Core Metadata specification's Schema section. I've made a few small changes in the Storage Format section (replacing JSON schemas with abstract representations) but haven't done a language pass on that section yet, and (per earlier discussion) we may want to move that section to the 3D Tiles or glTF Extension specs.

Changes:

- Language / clarity review.
- Replace JSON schema with tables and outlines
- Updates to account for latest `EXT_feature_metadata` changes

Despite removing the JSON schema I have tried to use normative names like `componentType` here; I expect the specification would be confusing without using such terms consistently. Using natural language like "Component Type" would probably also be fine.

donmccurdy commented:
/cc @ptrgags @lilleyse note that this PR is aganist the `3d-tiles-next` branch — should we create a similar `3d-tiles-next-rev` branch instead?

ptrgags commented:
@donmccurdy yes, I created a `3d-tiles-next-rev` branch for you. I already updated this PR to point to it.

donmccurdy commented:
In EXT_feature_metadata a schema has a "version" property, but I don't see that mentioned here. Should that be defined in this document, or is it outdated?

ptrgags commented:
ah, that `version` field is for if an application wants to version their external schema. 

Probably would be good to mention in this document

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
+| UINT32  | Unsigned integer in the range `[0, 4294967295]`                           |
+| INT64   | Signed integer in the range `[-9223372036854775808, 9223372036854775807]` |
+| UINT64  | Unsigned integer in the range `[0, 18446744073709551615]`                 |
+| FLOAT32 | A number that can be represented as 32-bit IEEE floating point            |
+| FLOAT64 | A number that can be represented as 64-bit IEEE floating point            |
```
we probably should be clearer that `NaN` and `Infinity` are excluded

donmccurdy replied:
Added the restriction just below the table. 👍

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
+Properties representing numeric values, fixed-length numeric arrays, and vectors may specify a minimum (`minimum`) and maximum (`maximum`). Minimum and maximum values may represent bounds of the valid range for a property, or the exact minimum and maximum values found in the dataset.
 
-Numeric properties may specify a minimum and maximum allowed value. For example a property storing GPS coordinates would define a range of `[-180, 180]` degrees for longitude values and `[-90, 90]` degrees for latitude values.
+> **Example:** A property storing GPS coordinates might define a range of `[-180, 180]` degrees for longitude values and `[-90, 90]` degrees for latitude values. If the dataset happens to contain only GPS coordinates in a small region, more specific ranges may be given instead.
```
```suggestion
> **Example:** A property storing GPS coordinates might define a range of `[-180, 180]` degrees for longitude values and `[-90, 90]` degrees for latitude values. If the dataset contains GPS coordinates only in a small region, more specific ranges may be given instead.
```

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
+> | property  | componentType | required | noData    |
+> |:----------|:--------------|:---------|:----------|
+> | height    | "FLOAT32"     | ✓        |           |
+> | species   | "ENUM"        |          | "Unknown" |
```
Oh in the paragraph above you mention array types but not enums (which here use the name, rather than the integer value.

donmccurdy replied:
Done ✅

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -372,64 +316,40 @@ This example shows how to encode a variable-length array of integers.
+- class `"weather"`
+- properties
+  - `"forecastHistory"`
+    - buffer: `[R, a, i, n, ...]`
```
Maybe be more explicit that the buffer stores the individual characters? not sure the best way to indicate this, here's one idea:

```suggestion
    - buffer: `['R', 'a', 'i', 'n', ...]`
```

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -320,18 +264,18 @@ Property vales are encoded in binary according to their type, in little-endian f
+| UINT32  | 32-bit unsigned integer                |
+| INT64   | 64-bit two's complement signed integer |
+| UINT64  | 64-bit unsigned integer                |
+| FLOAT32 | 32-bit IEEE floating point             |
+| FLOAT64 | 64-bit IEEE floating point             |
```
```suggestion
| FLOAT32 | 32-bit IEEE floating point number      |
| FLOAT64 | 64-bit IEEE floating point number      |
```

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
 
-The two formats are defined:
+* **Table format** - property values are stored in parallel 1D arrays
+* **Raster format** - property values are stored in channels of a 2D grid of pixels
```
I think we briefly talked about this, could we word this in a way that generalizes to 3D or N-D rasters?

donmccurdy replied:
Rephrased: 

* **Raster format** - property values are stored in channels of pixel-based formats (usually images)

... The raster format is for storing fine-grained metadata in pixel-based formats such as images and video.

ptrgags reviewed (commented):
@donmccurdy looks like this is heading in the right direction. Left some comments, mostly about wording or other proofreading things.

donmccurdy commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
+The `ARRAY` type is used to define a fixed- or variable-length array of components. For fixed-length arrays, a component count denotes the number of components in each array, and must be ≥2. Variable-length arrays do not define a component count, and arrays may have any length, including zero.
 
-The following enum value types are supported: `INT8`, `UINT8`, `INT16`, `UINT16`, `INT32`, `UINT32`, `INT64`, and `UINT64`. See [Property Types](#property-types).
+The `VECN` and `MATN` types represent specific subsets of the fixed-length `ARRAY` type, where `VECN` is a vector with `N` numeric components and `MATN` is an `N x N` matrix with `N²` numeric components. Where applicable, authoring implementations should choose these more specific types to improve data portability, particularly for mathematical types. Schema representations may choose to make com ...
```
Great point on column-major order, definitely want to include that!

I'll just slim down the "data portability" sentence to the line below; I don't think it's necessary to list benefits here.

> Where applicable, authoring implementations should choose these more specific types.

donmccurdy commented:
@ptrgags I've addressed earlier feedback, and added `schema.version` to the document. Wasn't quite sure how to describe it beyond "application-specific string", is this reasonable?

> ﻿﻿Schema version (`version`) is an application-specific identifier for a given schema revision.

We could consider requiring it to be a valid [semver string](https://semver.org/), although definitions of major/minor/patch changes are not as obvious for file formats as for software. But this at least provides a well-defined way to tell if one string is "newer" than another.

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
 ## Overview
 
-The Cesium 3D Metadata Specification defines a standard metadata format for 3D data. This spans a myriad of domains, such as heights of buildings in a city, names of different structures in a CAD model, material properties for textured surfaces, and classification codes for point clouds.
+The Cesium 3D Metadata Specification defines a standard format for structured metadata in 3D content. Complementary to general-purpose standards like XMP, this specification allows metadata — represented as entities and properties — to be closely associated with parts of 3D content, with data representations appropriate for large, distributed datasets. For the most detailed use cases, properties  ...
```
Yeah could you make a small PR for this in the glTF extension so we don't forget about it?

donmccurdy commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
+Descriptions (`description`) provide a human-readable explanation of a property, its purpose, or its contents. Typically at least a phrase, and possibly several sentences or paragraphs. To provide a machine-readable semantic meaning, a property must also define a [semantic](#semantics).
 
-An array property must declare a component type, which is the type of each component in the array. A component type may be any type from the table above with the exception of `ARRAY`; arrays of arrays are not supported.
+#### Semantic
```
Loosened, making this a recommendation only.

donmccurdy commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
+Property IDs, names, and descriptions do not imply semantic meaning within the data format. To provide such a meaning, properties may be assigned a semantic (`semantic`). A semantic is an identifier describing how the property's content should be interpreted, according to an external or application-specific semantic specification.
 
-Arrays can be used to represent vector and matrix properties as shown in the example below.
+> **Example:** Common semantic definitions might include temperature in degrees Celsius, time in milliseconds, or mean squared error (MSE).
```
I've added a link to `./Semantics/README.md` in the paragraph above – it's part of the same specification, so probably no need for a subset here?

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
+The Cesium 3D Metadata Specification defines a standard format for structured metadata in 3D content. Complementary to general-purpose standards like XMP, this specification allows metadata — represented as entities and properties — to be closely associated with parts of 3D content, with data representations appropriate for large, distributed datasets. For the most detailed use cases, properties  ...
 
-This specification defines a set of core concepts to be used by multiple 3D formats. It is language and format independent. This document occasionally uses JSON examples for illustrating concepts but does not define a JSON schema.
+Many domains benefit from structured metadata — typical examples include historical details of buildings in a city, names of components in a CAD model, descriptions of regions on textured surfaces, and classification codes for point clouds.
```
We discussed this earlier this week, we settled on only doing the rename in the glTF extension.

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,227 @@ Draft
+**Entities** are instantiations of class, populated with **property values** conforming to the class definition. Every property value of an entity must be defined by its class, and an entity must not have extraneous property values. Properties of a class may be required, in which case all entities instantiating the class are required to include them.
 
-A schema describes what properties are available. However, it does not describe how property values are stored. This allows the schema to be defined separately from the data itself. This has several benefits:
+>  **Implementation note:** Entities may be defined at various levels of abstraction. Within a large dataset, individual vertices or texels may represent entities with granular metadata properties. The entire dataset may itself be an entity of another class, with its own associated properties.
```
Yes, I like this alternate wording, please make this change.

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,240 @@ Draft
 - [Overview](#overview)
```
Please add yourself to the contributors list above between Sam and Bao, thanks!

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -21,283 +21,240 @@ Draft
+#### Semantic
 
-Enum properties are denoted by the `ENUM` type. An enum property must provide the ID of the specific enum it uses.
+Property IDs, names, and descriptions do not impute meaning. To provide such a meaning, properties may be assigned a semantic identifier string (`semantic`), indicating how the property's content should be interpreted. Semantic identifiers may be defined by the [Cesium Metadata Semantic Reference](./Semantics/) or by external semantic references, and may be application-specific. Identifiers shoul ...
```
Comparing the Semantics link to the rest of the document, it looks like we've been inconsistent with the links. I _think_ it's `./Semantics/` like you have here, but good to double check.

ptrgags reviewed (commented):
@donmccurdy changes are looking good, just a couple small comments here

donmccurdy commented:
@ptrgags thanks! Have caught up on feedback for this PR now and fixed a number of broken links.

ptrgags commented on specification/Metadata/README.md:
```diff
@@ -94,7 +95,7 @@ Schema version (`version`) is an application-specific identifier for a given sch
 ### Enums
 
-An enum consists of a set of named values, represented as `string: integer` pairs. The following enum value types are supported: `INT8`, `UINT8`, `INT16`, `UINT16`, `INT32`, `UINT32`, `INT64`, and `UINT64`. See [Property Types](#property-types) for definitions of each. Smaller enum types limit the range of possible enum values, and allow more efficient binary encoding. For unsigned value types, e ...
+An enum consists of a set of named values, represented as `string: integer` pairs. The following enum value types are supported: `INT8`, `UINT8`, `INT16`, `UINT16`, `INT32`, `UINT32`, `INT64`, and `UINT64`. See [Property Types](#type) for definitions of each. Smaller enum types limit the range of possible enum values, and allow more efficient binary encoding. For unsigned value types, enum values ...
```
I would also change the name of the link to be closer to the section heading for clarity
```suggestion
An enum consists of a set of named values, represented as `string: integer` pairs. The following enum value types are supported: `INT8`, `UINT8`, `INT16`, `UINT16`, `INT32`, `UINT32`, `INT64`, and `UINT64`. See the [Type](#type) section for definitions of each. Smaller enum types limit the range of possible enum values, and allow more efficient binary encoding. For unsigned value types, enum values most be non-negative. Duplicate names or values within the same enum are not allowed.
```

ptrgags reviewed (approved):
Looks good, I'm going to make a tiny tweak to one link title and then merge

ptrgags commented:
Thanks @donmccurdy!
